src/runtime_src/core/edge/user/system_linux.cpp

```diff
@@ -43,7 +50,158 @@
+// [int, device type] -> [device index, shim/driver type]
+// As either of AIARM or ZOCL or both can be active together
+// we need mapping of device id to device type
 namespace {
+enum class dev_type : uint16_t
+{
+  aiarm = 0,
+  zocl = 1
+};
+static std::unordered_map<uint16_t, dev_type> dev_map;
+
+static unsigned int
+enumerate_devices()
+{
+  /*
+   * Zocl node is not present in some platforms static dtb, it gets loaded
+   * using overlay dtb, drm device node is not created until zocl is present
+   * So if enable_flat is set return 1 valid device
+   * Valid for SOM kind of platforms
+   */
+  if (xrt_core::config::get_enable_flat())
+    return 1;
+
+  static const std::string render_dev_sym_dir{"/dev/dri/by-path/"};
+  std::string aiarm_render_devname;
+  std::string zocl_render_devname;
+
+  // On Edge VE2 kind of devices 'telluride_drm' is the name of aiarm node in device tree
+  // On Edge traditional devices 'zyxclmm_drm' is the name of zocl node in device tree
+  // A symlink to render devices are created based on these node names
+  try {
+    static const std::regex aiarm_filter{"platform.*telluride_drm-render"};
+    static const std::regex zocl_filter{"platform.*zyxclmm_drm-render"};
+    if (!std::filesystem::exists(render_dev_sym_dir)) {
+      std::string msg{"DRM device path - " + render_dev_sym_dir +
+                      "doesn't exist, cannot detect devices"};
+      xrt_core::message::send(xrt_core::message::severity_level::error, "XRT", msg);
+      return 0;
+    }
+    std::filesystem::directory_iterator end_itr;
+    for (std::filesystem::directory_iterator itr{render_dev_sym_dir}; itr != end_itr; ++itr) {
+      if (std::regex_match(itr->path().filename().string(), aiarm_filter)) {
+        aiarm_render_devname = std::filesystem::read_symlink(itr->path()).filename().string();
+      }
+      if (std::regex_match(itr->path().filename().string(), zocl_filter)) {
+        zocl_render_devname = std::filesystem::read_symlink(itr->path()).filename().string();
+      }
+    }
+  }
+  catch (const std::exception& e) {
+    xrt_core::message::send(xrt_core::message::severity_level::error, "XRT",
+                            std::string{"Unable to read renderD* path of devices"} + e.what());
+    return 0;
+  }
+
+  unsigned device_count = 0;
+  // lambda function that checks validity of /dev/dri/renderD* device node
+  // and inserts in global map with its type of device and increments device count
+  auto validate_and_insert_in_map =
+      [&](const std::string& drm_dev_name, const std::string& ver_name, const enum dev_type& type) {
+    try {
+      if (!std::filesystem::exists(drm_dev_name))
+        throw std::runtime_error(drm_dev_name + " device node doesn't exist");
+
+      auto fd_close = [](int* fd){
+        if (fd && *fd >= 0) {
+          close(*fd);
+          delete fd;
+        }
+      };
+      auto fd = std::unique_ptr<int, decltype(fd_close)>(new int(open(drm_dev_name.c_str(), O_RDWR)), fd_close);
```

Comment:
The extra new is kind of ugly.  I would have done this:
```
auto fd_close = [](int* fd) {
  if (fd && *fd >= 0)
    close(*fd);
};

int main()
{
  auto fd = open();
  std::unique_ptr<int, decltype(fd_close)> (&fd, fd_close);
  return 0;
}
```

Reply:
Yes Soren, I agree this eliminates new. Made changes accordingly. Please review